Support for scene debugger

keianhzo opened the pull request:
Scene debugging is a tedious process and involves:
- Creating a scene with the appropriate flags (and knowing those flags)
- In some cases you need to accept certificates
- Exporting the scene manually
- Drag&drop the GLB in the room

This PR adds a `Create` and `Update` buttons to the tools panel to simplify the room creation and the room scene update process.

This PR depends on https://github.com/mozilla/hubs/pull/6383

Exairnous reviewed (changes requested):
This is just an initial portion of my review, I will post more later.

Installing the dependencies seems prone to breakage depending on Blender/Python version, e.g. I got this when trying to install the dependencies in Blender 2.93
`/home/user/Workshop/CustomApps/blender-2.93.17/2.93/python/bin/python3.9: No module named pip`.
Blender 2.93 uses Python 3.9 while later Blenders use Python 3.10.  Also, no indication was given to the user that installing the dependency had failed.

Trying to install it in Blender 3.6 worked better, but produced this warning at the end:
```
Installing collected packages: sortedcontainers, urllib3, sniffio, pysocks, idna, h11, exceptiongroup, certifi, attrs, wsproto, outcome, trio, trio-websocket, selenium
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
requests 2.27.1 requires urllib3<1.27,>=1.21.1, but you have urllib3 2.1.0 which is incompatible.
Successfully installed attrs-23.1.0 certifi-2023.11.17 exceptiongroup-1.2.0 h11-0.14.0 idna-3.4 outcome-1.3.0.post0 pysocks-1.7.1 selenium-4.15.2 sniffio-1.3.0 sortedcontainers-2.4.0 trio-0.23.1 trio-websocket-0.11.1 urllib3-2.1.0 wsproto-1.2.0
```

Continuing on, I tried to create a room, and that worked fine (once I'd switched it from the default hubs.local instance URL.  I think this should default to the demo instance to make it easier for people).

Then when I tried to update the room it errored out with an error about not being able to find the exported GLB file:
```
11:01:40 | INFO: Draco mesh compression is available, use library at /home/user/Workshop/CustomApps/blender-3.6.5/3.6/python/lib/python3.10/site-packages/libextern_draco.so
[Errno 13] Permission denied: '/Users'
Traceback (most recent call last):
  File "/home/user/Workshop/Programming/Blender/user_scripts_dir/addons/io_hubs_addon/components/ui.py", line 235, in execute
    refresh_scene_viewer()
  File "/home/user/Workshop/Programming/Blender/user_scripts_dir/addons/io_hubs_addon/components/ui.py", line 212, in refresh_scene_viewer
    file_input.send_keys(os.path.join(bpy.app.tempdir, EXPORT_TMP_FILE_NAME))
  File "/home/user/.local/lib/python3.10/site-packages/selenium/webdriver/remote/webelement.py", line 230, in send_keys
    self._execute(
  File "/home/user/.local/lib/python3.10/site-packages/selenium/webdriver/remote/webelement.py", line 394, in _execute
    return self._parent.execute(command, params)
  File "/home/user/.local/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py", line 347, in execute
    self.error_handler.check_response(response)
  File "/home/user/.local/lib/python3.10/site-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidArgumentException: Message: File not found: /tmp/blender_fVu4ln/__hubs_tmp_scene_.glb
Stacktrace:
RemoteError@chrome://remote/content/shared/RemoteError.sys.mjs:8:8
WebDriverError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:187:5
InvalidArgumentError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:378:5
interaction.uploadFiles@chrome://remote/content/marionette/interaction.sys.mjs:543:13

Error: Python: Traceback (most recent call last):
  File "/home/user/Workshop/Programming/Blender/user_scripts_dir/addons/io_hubs_addon/components/ui.py", line 235, in execute
    refresh_scene_viewer()
  File "/home/user/Workshop/Programming/Blender/user_scripts_dir/addons/io_hubs_addon/components/ui.py", line 212, in refresh_scene_viewer
    file_input.send_keys(os.path.join(bpy.app.tempdir, EXPORT_TMP_FILE_NAME))
  File "/home/user/.local/lib/python3.10/site-packages/selenium/webdriver/remote/webelement.py", line 230, in send_keys
    self._execute(
  File "/home/user/.local/lib/python3.10/site-packages/selenium/webdriver/remote/webelement.py", line 394, in _execute
    return self._parent.execute(command, params)
  File "/home/user/.local/lib/python3.10/site-packages/selenium/webdriver/remote/webdriver.py", line 347, in execute
    self.error_handler.check_response(response)
  File "/home/user/.local/lib/python3.10/site-packages/selenium/webdriver/remote/errorhandler.py", line 229, in check_response
    raise exception_class(message, screen, stacktrace)
selenium.common.exceptions.InvalidArgumentException: Message: File not found: /tmp/blender_fVu4ln/__hubs_tmp_scene_.glb
Stacktrace:
RemoteError@chrome://remote/content/shared/RemoteError.sys.mjs:8:8
WebDriverError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:187:5
InvalidArgumentError@chrome://remote/content/shared/webdriver/Errors.sys.mjs:378:5
interaction.uploadFiles@chrome://remote/content/marionette/interaction.sys.mjs:543:13
```

When I look in Blender's temp directory the GLB file indeed isn't there, but if I manually paste the relevant code from the `export_scene` function into Blender's Python console it works fine and the file gets created in the temp directory, so I'm not sure why it's not working.

Exairnous commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -142,6 +145,271 @@ def draw(self, context):
+            'export_apply': True
+        }
+        bpy.ops.export_scene.gltf(**args)
+    except Exception as err:
+        print(err, file=sys.stderr)
```
I don't think we should just print the error and continue on from this, if the glb export fails then the update process should be stopped and the user notified.

Exairnous replied:
Thanks!  Yeah, the `hubs_report_viewer` does a good job :)

Exairnous commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -142,6 +145,271 @@ def draw(self, context):
+
+                return {'FINISHED'}
+
+        except Exception as e:
+            print(e)
+            return {"CANCELLED"}
```
What issues are being guarded against here?  Also, unless this is one of those cases where the same thing multiple times in a row makes it work, the user should probably be notified with a descriptive error message (or even if it does it might be nice to display a message that tells the user to try again).

keianhzo replied:
I've added the error report so that should be enough.

Exairnous replied:
Seems to be good.  Thanks.

Exairnous reviewed (changes requested):
After your most recent changes I was able to get it to work and it's predominantly working quite nicely.  However, I've left some inline comments and I have a a few more general thoughts and comments I'll list here.

* The scene debugger code is all mixed in with the UI code, but it should really be in its own separate file. The `HUBS_PT_ToolsPanel` should stay though as it's a generic UI element.
* Can an existing room be used?  Or can a lighter scene be specified for room creation?  Currently it takes a long time for the room to load for me for the first time because it's having to download a large scene (at least on the demo server) that I'm going to immediately overwrite.
* The install dependencies operator doesn't provide any feedback if the install fails.  I think it can be really easy to miss that it's failed, especially if it fails early in the process and would benefit from an error message being reported to the user.
* The switch window/bring to front code isn't working for Firefox for me, but it works great in Chromium.  Since the browser isn't using my window manager decorations I also have no option to keep it always on top to work around the issue.

keianhzo commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -142,6 +145,271 @@ def draw(self, context):
+        if hubs_store:
+            import json
+            hubs_store = json.loads(storage.get("___hubs_store"))
+            has_credentials = "credentials" in hubs_store
```
Yeah, I just saw that. In my case the keys can exist but they be null so we need to account for that case too. Updated.

Exairnous replied:
Thank you.  Works for me now as well.

keianhzo commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -142,6 +145,271 @@ def draw(self, context):
+        import os
+        import sys
+        extension = '.glb'
+        output_dir = bpy.app.tempdir
+        if not os.path.exists(output_dir):
+            os.makedirs(output_dir)
```
Oh yeah. Left over from a previous version. we weren't even using that output_dir variable.

keianhzo commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -142,6 +145,271 @@ def draw(self, context):
+            # Settings from "Remember Export Settings"
+            **dict(bpy.context.scene.get('glTF2ExportSettings', {})),
+
+            'export_format': ('GLB' if extension == '.glb' else 'GLTF_SEPARATE'),
+            'filepath': os.path.join(bpy.app.tempdir, EXPORT_TMP_FILE_NAME),
+            'export_cameras': True,
+            'export_lights': True,
+            'export_extras': True,
+            'use_visible': True,
+            'export_apply': True
```
Actually this is also left over from a previous version. My idea was to totally rely on the exporter settings. At least for the initial version. After that we can look for a way to maybe expose the most common export settings in the Hubs panel to override the glTF exporter settings if explicitly requested by the user. For now I'll remove this overrides.

Exairnous replied:
I did some checking into embedding the glTF export panel and sadly I don't think it's an option.  I don't know if maybe we want to have a toggle between overriding settings and just using the saved glTF properties, but I think if we're going to have overrides then it would be nice to have all the `Limit to` options (although that depends on Blender version).  Aside from those, the `Rest & Ranges` options could affect scene preview, and I think `Sampling Animations` should be forced off (have it there, but greyed out and disabled, or maybe just a label that says sample animations disabled?) because it can cause problems with animations and increases the file size.

I'm not really sure what to say about this to be honest.  I don't think anything will fit perfectly, and what you have currently, technically, allows advanced users to squeak by and accomplish all possible exports, but it's not super intuitive.

Exairnous replied:
This is generally good, thanks, but I think the fact we are forcing no sampling of animations should be conveyed to the user, along with our reasoning (in a tooltip I guess), especially since it defaults to on in the regular glTF export.  Also, the use_active_scene property was only introduced in Blender 3.2, but yes,having use_active_scene on is useful, but again I think we want to let the user know that because if they do have multiple scenes, but don't know to check that button in the exporter, then I believe it will work for the debugger, but not when they export to upload to Spoke (because they didn't know to check that export option).

keianhzo commented on addons/io_hubs_addon/preferences.py:
```diff
@@ -25,6 +27,102 @@ def get_recast_lib_path():
+        import subprocess
+        import sys
+
+        subprocess.run([sys.executable, '-m', 'pip', 'install', '--upgrade', 'pip'],
+                       capture_output=False, text=True, input="y")
```
Probably not, I added that pretty early but I don't remember why... I have removed it for now soI guess we will find out.

Exairnous replied:
Thanks.  I think this should be good.  But the ensure pip line needs to be called in the install dep operator (adding it there fixed the module not getting installed in 2.93 for me) as well as in the uninstall operator.

keianhzo replied:
> Thanks. I think this should be good. But the ensure pip line needs to be called in the install dep operator (adding it there fixed the module not getting installed in 2.93 for me) as well as in the uninstall operator.
Humm, ok, I'll add that.

> I also noticed when I got it working on 2.93 that the isModuleAvailable function didn't work until after I'd restarted Blender (like you mentioned in the add-on meeting today), but further uninstalls and reinstalls haven't triggered this, so it seems to be intermittent.

Yeah, I saw that but I don't know why either. I'll keep an eye on it.

keianhzo commented on addons/io_hubs_addon/preferences.py:
```diff
@@ -25,6 +27,102 @@ def get_recast_lib_path():
+                [name for name, _ in self.dep_names.items()]],
+            capture_output=False, text=True, input="y")
+
+        import os
+        from .utils import get_user_python_path
+        selenium_path = os.path.join(get_user_python_path(), "selenium")
+        if self.force and os.path.isfile(selenium_path):
+            import shutil
+            shutil.rmtree(selenium_path)
```
Good call, updated.

Exairnous replied:
Thank you!

keianhzo commented:
@Exairnous I think I've fixed most of the feedback. I've also added a few export flags. Let me know what you think.

keianhzo commented:
I've added a few more improvements:
- Visual cue for room status
- Close room button
- I've moved the instance management to the Hubs panel and I have added support for multiple server instances management. I think that can be useful to switch between instances.

I haven't found a good way of starting a room with an empty scene, that might be something to add later, anyways it's only the first time you create it after that it should already be overriden.

keianhzo commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -142,6 +145,271 @@ def draw(self, context):
+                    params = f'{params}&vr_entry_type=2d_now'
+
+                web_driver.get(
+                    f'{get_addon_pref(context).hubs_instance_url}?{params}')
```
I've added support for multiple instances and also now the demo instance is the default one when you create an instance.

Exairnous replied:
I consistently get selenium complaining that it's a malformed URL if I omit the `https://` in my instance/room URLs.  Not sure why.

keianhzo replied:
Interesting, I'm seeing that too. I'm going to force the protocol in the URL.

Exairnous commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -142,6 +145,271 @@ def draw(self, context):
+                     )
+        top_row.label(text="Scene Debugger")
+
+        if context.scene.hubs_scene_debugger_expanded:
```
You're welcome!

Exairnous commented on addons/io_hubs_addon/components/ui.py:
```diff
@@ -229,7 +516,18 @@ def unregister():
+
+    global web_driver
+    if web_driver and isWebdriverAlive(web_driver):
+        web_driver.close()
```
Well, I did some more checking and it turns out I was working on outdated information.  It turns out that Blender has called unregister on exit since versions 2.91.0 and 2.83.9[1][2], and my own tests confirm this.  But, you say you think you saw a few instances where it wasn't being called, so now I'm even more curious.  Do you remember which exit situations didn't call unregister?

[1] https://projects.blender.org/blender/blender/commit/fa566157a5c351775d082b05b180c630665b4afc)
[2] https://blender.stackexchange.com/questions/153084/blender-add-on-dev-unregister-is-never-called-on-blender-exit

keianhzo replied:
I can't, I'm going to remove it from now, we can always add it it we find a good reason.

Exairnous commented on addons/io_hubs_addon/preferences.py:
```diff
@@ -25,6 +27,102 @@ def get_recast_lib_path():
+class DeleteProfileOperator(bpy.types.Operator):
+    bl_idname = "pref.hubs_prefs_remove_profile"
+    bl_label = "Delete"
+    bl_description = "Delete Browser profile"
```
Thank you!

Exairnous reviewed (changes requested):
Thanks.  This is looking good.  I've left some ideas about the animation export options in one of the inline comments, but pretty much all of them can affect the export to a significant degree.  I don't know how many we want to add and since there is the option of using the saved glTF settings... as I said in the inline comment, I'm not sure what the best course of action is here.

Aside from the inline comments, here are some more general thoughts.

There's an unused import (`bpy.utils.previews`) in components_registry.py

The instances/rooms shouldn't be saved to the preferences as this goes against Blender's usual behavior.  Generally, stuff in the N-Panel affects only stuff saved in the current blend file, and while I understand wanting this to be global, aside from it being unexpected behavior, saving it to the preferences interferes with setups that don't auto-save the preferences, plus I have noticed occasions where the data hasn't been saved properly and is lost on restart, and if you remove and then reinstall the add-on (as is done in an upgrade) you will lose the user preferences/instance/room data.  So all in all I would recommend storing the instances and rooms in the scene.  

The user can still add some instances/rooms and then save the startup.blend to make it almost global and this is Blender's standard way of working with stuff like this (in my experience, at least).

Also, while it's certainly not bad to have the room flags global, I think there might be a benefit to storing the room flags with each room/instance url (and modifying or picking from that when opening/creating a room) because I think I'm more likely to want to use the same flags with the same instances/rooms, e.g. for the behavior graphs I'll always want to use the new loader and physics debug flag (if that gets added as you mentioned at the add-on meeting), while for the demo instance I'll more likely want to use the old loader and no debugging, so it might be nice to be able to just choose the saved room/instance and have the flags all ready to go.  But I'm not entirely sure on this, and saving the room/instance data with the blend file may provide enough flexibility that this isn't needed, not sure.

I think it would be nice if the room name of the open room was displayed in the top status area e.g. `In Room: Wonderful Laughing Duck`.  Another benefit of keeping track of the current room is that it would allow the open room button to change to say something like refresh room when you select the currently open room from the list.

I don't think the demo instances list needs to be greyed out when a room is open, the create room button yes, but I don't think there's any harm in leaving the rest unlocked so you can add/remove instances even with a window open.

Also, as you said at the add-on meeting, I don't think sign in is necessary if you aren't using debugLocalScene, so skipping straight to the In Room status makes sense to me there.

Finally, in the past, pressing the update button has failed to update the scene sometimes (the glb gets exported, but doesn't seem to make it to the room), however in my most recent batch of testing I have been unable to trigger it, so I guess we keep an eye out, ebut proceed as normal.

Oh, and I talked to Jim and he said that neither of his browsers were coming to the foreground.  Also, Firefox still isn't for me, but Chromium does.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -0,0 +1,651 @@
+    bl_parent_id = "HUBS_PT_ToolsSceneDebuggerPanel"
+
+    def draw(self, context: Context):
+        if isModuleAvailable("selenium"):
```
Nice, sdded.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -0,0 +1,651 @@
+    bl_description = "Create room"
+    bl_options = {'REGISTER', 'UNDO'}
+
+    url: bpy.props.StringProperty()
```
yup.

keianhzo commented on addons/io_hubs_addon/debugger.py:
```diff
@@ -0,0 +1,651 @@
+    def execute(self, context):
+        prefs = get_addon_pref(context)
+        prefs.hubs_instances.remove(prefs.hubs_instance_idx)
+        prefs.hubs_instance_idx = len(prefs.hubs_instances) - 1
```
I think this should work too, is it failing for you?

keianhzo replied:
Fixed

GottfriedHofmann commented:
> > The instances/rooms shouldn't be saved to the preferences as this goes against Blender's usual behavior. Generally, stuff in the N-Panel affects only stuff saved in the current blend file, and while I understand wanting this to be global, aside from it being unexpected behavior, [...]
> Is there any information about how it goes about Blender's usual behavior? I understand that in most cases you might not necessarily want this behavior but in this specific case I find it useful as you can reuse instances/rooms in any blender scene that work on without needing to copy&paste. Specially for instances you are probably going to always use the same ones in every blender scene.

I'm also interested in that information. The sidebar (or N-Panel) is the place to go for addons in general. Having global settings there is not unexpected, at least for users who have experience with addons. I'm working on another addon where we solved this by having global settings exposed in both the addon preferences and the sidebar.

keianhzo commented:
Thanks for the feedback @GottfriedHofmann. We have ended up moving this to the filesystem as a JSON file.

Exairnous reviewed (changes requested):
You're welcome!  Thank you for the changes.  There's still a couple minor things I've left comments about, but everything's looking pretty good now.
> Is there any information about how it goes about Blender's usual behavior? I understand that in most cases you might not necessarily want this behavior but in this specific case I find it useful as you can reuse instances/rooms in any blender scene that work on without needing to copy&paste. Specially for instances you are probably going to always use the same ones in every blender scene.
> 
> I'm going to ask the community what's the preferred behavior because I'm not really sure.

Thank you for putting this out to the community, I've put my reasoning in my reply to Gottfried Hofmann below.

> My approach to this was that the room flags and the export flags are stored per room and the instances/rooms are stored in prefs because you are likely to reuse instances/rooms across scenes but you probably use the different flags in different scenes. I can imagine that while working on a room you won't change the rooms flags that much but if you need to, you can just check/uncheck and try just clicking on update Vs the other way around you'll need to create a new room and save or manually tweak the url which not everyone knows how to do.

Ah, my apologies, I didn't convey that idea very well.  What I had meant when I said to store the flags with the URLs was to store them as boolean properties in the property group, not as query string parameters in the URL.  But you're right that you won't be changing the room flags that much and that they'll probably be specific to the scene, so keeping it simple like you have it currently is better.  Please ignore this suggestion :)

> Yes, I think this is based on some initial limitation that I found but it does not make sense anyomore. I've updated so all add/remove instance/room and create/open room are always enabled.

Nice.  But now that you can create rooms with a browser opened, I noticed that it isn't being brought to the foreground (open room also doesn't).  So, do we want the browser to come to the foreground in these cases?

> It seems to be some sort of selenium bug, I've found a hack, not great but it's better than a broken functionality.

Excellent!  This is working nicely for me now.

> We have ended up moving this to the filesystem as a JSON file.

Yes, I like this much better, thank you.  However, There doesn't appear to be a default instance anymore, and while not really necessary, I thought it was a nice touch.

Also, now that there are three Hubs things being saved to the home directory, what do you think about grouping them in a generic Hubs folder?  It's not a big deal either way, but it might be nice.

Exairnous commented:
Hi @GottfriedHofmann , thank you for your interest.  You're right, of course, that the sidebar/N-Panel is the main go to for add-ons, and that having these as global settings is desirable, but there were a couple reasons why I didn't like the implementation as it was.  Let me try to explain my reasoning better, when I use the things in the N-Panel I find they usually act on data in the current scene, e.g. object properties, tool settings, view options, and even for the add-ons that I use (mostly bundled add-ons or my own, but a few third party ones like TexTools as well) they also operate on data in the current scene.  While there are some that I know of that do operate on preferences, like I think there was an add-on for manipulating studio lights, if I recall correctly that was the only way the studio lights add-on could operate and what it displayed was mirrored in the preferences like your own add-on.  The difference here is that while the settings were saved to the preferences they weren't displayed in the preferences and no indication was given that they weren't operating on the scene data like other options in the same panel were (e.g. the room flags displayed just above are saved to scene data, not the preferences).  Aside from that, creating rooms introduces volatile data (the rooms are created on a server and the only reference to them is saved to the add-on settings, unless you favorite them or bookmark them, etc., but I expect users would just save them to the add-on settings for this case), and while not a huge loss, having this wiped on updating the add-on is more destructive than when you just have to re-enter data saved in another place (also, for some reason the settings sometimes weren't getting saved, but that probably could have been fixed).  Finally, (and this was probably my biggest reason why I didn't like them being saved to the preferences) the preferences were being auto-saved on every modification with no indication to the user this was happening, so for example, if I don't have my preferences set to auto save (which I don't) and I disable the Load UI option for a blend file (which I do occasionally), and then I create a room and add it to my list of saved rooms, then the next time I start up Blender and load a blend file the UI won't load as expected because the setting I thought was reverted when I closed Blender was saved without my knowledge (and this could happen with a bunch of settings over time, silently screwing up your preferences).  If Blender allowed you to update specific settings only and there wasn't the danger of losing the data on add-on updates, then I wouldn't object to storing the instance and room lists there.

So that's why I initially suggested switching to storing the settings in the scene data, but that was less than ideal because it wasn't global, and then people raised the problem that sensitive data could be leaked when sharing blend files, so after a bit of back and forth we decided on storing them in a JSON file in your home directory which solves all the above problems and adds the benefit of being Blender version agnostic.

If you have any further input on this you're more than welcome to share it.  You probably know more about the standard behavior of add-ons in cases like this than I do.

keianhzo commented:
I think I've addressed all feedback and this is ready for shipping.

> Nice. But now that you can create rooms with a browser opened, I noticed that it isn't being brought to the foreground (open room also doesn't). So, do we want the browser to come to the foreground in these cases?

Oh yep, I've added that so we always bring to front if not alive.

>Yes, I like this much better, thank you. However, There doesn't appear to be a default instance anymore, and while not really necessary, I thought it was a nice touch.

That must have get lost while migrating the preferences. Re-added.

>Also, now that there are three Hubs things being saved to the home directory, what do you think about grouping them in a generic Hubs folder? It's not a big deal either way, but it might be nice.

Yeah, I've done some reorg to better fit all he new fs files.

Exairnous commented:
Thank you, this is looking excellent!

> I think I've addressed all feedback and this is ready for shipping.

It's almost ready for shipping, but not _quite_.  I was finally able to find reproducible steps to trigger the GLB failing to drop in-room:
1. Set your browser to Firefox. (I'm not sure if this has any impact, but it's what I was using)
2. Uncheck the New Loader and Debug Local Scene options (leave the other options checked).
3. Create a room on the Demo Hub (or open a previously saved room on the Demo Hub).
4. Press the update scene button.
5. The browser comes to the front, but the object isn't dropped into the scene.

Aside from this, I think you didn't answer two of my inline comments?
Re removing rooms/instances:
>It isn't failing, but it always jumps down to the end of the list when you remove an item. The code from the loop animation component keeps the selection in the same relative place in the list and using it here as well makes the behavior consistent across the add-on.

Re export options:
> This is generally good, thanks, but I think the fact we are forcing no sampling of animations should be conveyed to the user, along with our reasoning (in a tooltip I guess), especially since it defaults to on in the regular glTF export. Also, the use_active_scene property was only introduced in Blender 3.2, but yes,having use_active_scene on is useful, but again I think we want to let the user know that because if they do have multiple scenes, but don't know to check that button in the exporter, then I believe it will work for the debugger, but not when they export to upload to Spoke (because they didn't know to check that export option).

If you generally agree with these two comments, but don't want to spend the time on them I'm happy to implement them in a followup PR or something.

Finally, if people use debug local scene and then try to open the room without debug local scene they'll run into this bug: https://github.com/mozilla/hubs/issues/6179.  This isn't anything to do with this PR, except more people may run into it now, so if you see anyone getting stuck on it, just point them to the GitHub issue for the workaround.

keianhzo commented:
Thanks again. I fixed all the above.

> I'm sorry, it's almost ready for shipping, but not quite. I was finally able to find reproducible steps to trigger the GLB failing to drop in-room:

Fixed

> It isn't failing, but it always jumps down to the end of the list when you remove an item. The code from the loop animation component keeps the selection in the same relative place in the list and using it here as well makes the behavior consistent across the add-on.

Fixed

> If you generally agree with these two comments, but don't want to spend the time on them I'm happy to implement them in a followup PR or something.

Yes, better address it in a follow-up PR. Thanks

> Finally, if people use debug local scene and then try to open the room without debug local scene they'll run into this bug: https://github.com/mozilla/hubs/issues/6179. This isn't anything to do with this PR, except more people may run into it now, so if you see anyone getting stuck on it, just point them to the GitHub issue for the workaround.

I'll take a look and fix this next week before releasing this.

Exairnous reviewed (approved):
Alright, then I think it's good to land (once Jim's issue he reported in Discord, Blender hanging when hitting the close button with Chrome open on Windows, is resolved) and I'll post a followup PR.

Exairnous commented:
Good catch!  Also, I've just been testing now and it seems like exporting multiple scenes no longer breaks Hubs, so maybe we don't need to force exporting the active scene now.  Can you confirm?

Exairnous commented:
I've added the followup PR and will update that as needed to any changes here.

Exairnous commented:
Overall, looks good.  Blender already has an open link operator though `wm.url_open`, so unless there's a reason not to use it, I think it's probably better to stick with Blender's.
Nit: and the active room flags can be a little bit messy with extraneous commas and equals signs sometimes.
`

Exairnous commented:
Actually, the room flags having equals signs can cause the update scene vs spawn object check to return the wrong one because it's looking for the exact string of "debugLocalScene" in a list and not finding it.

Exairnous commented:
Thank you for the updates, but I have some accessibility concerns with the Is Active status indicators being red and green.  For example, here is a simulation of Green-Blind/Deuteranopia applied to the indicators (https://www.color-blindness.com/coblis-color-blindness-simulator/)

![blender-addon-color-blindness-check-2](https://github.com/MozillaReality/hubs-blender-exporter/assets/9795121/2635d0ed-318e-4cb4-9b62-1a854a9ccf43)
<details>
<summary>Original</summary>
<img src='https://github.com/MozillaReality/hubs-blender-exporter/assets/9795121/9bd385e4-06ad-4447-a71d-373889b8b198'>
</details>



I didn't mention anything with the other status indicators because they were never the only indications of status, but these are.


I also think the room flags should have the actual parameter in brackets at the end of the tooltip.  As it is, the actual parameters are only referenced in the Python variable name which is only visible if people have turned Python tooltips on (they're off by default) and even then the `Skip Entry` one doesn't include the `2d_now` value.

keianhzo commented:
>Thank you for the updates, but I have some accessibility concerns with the Is Active status indicators being red and green. For example, here is a simulation of Green-Blind/Deuteranopia applied to the indicators (https://www.color-blindness.com/coblis-color-blindness-simulator/)

Maybe we can add the text info to the right (On/Off)?

>I also think the room flags should have the actual parameter in brackets at the end of the tooltip. As it is, the actual parameters are only referenced in the Python variable name which is only visible if people have turned Python tooltips on (they're off by default) and even then the Skip Entry one doesn't include the 2d_now value.

Yeah I guess it wouldn't harm.

Let's open PRs/issues for this to track

Exairnous commented:
Adding (On/Off) to the right would make it too wide, I think.

I've opened up a couple PRs with possible improvements.  Let me know what you think.
#257 #258
